### ISSUE TITLE
Does not build on ARM

### ISSUE THREAD

**Issue by russoue** [facebook/watchman#302]
I tried to build it on a `Raspbian GNU/Linux 8 (jessie)` but am getting the following error:

    # russoue at raspberrypi in ~/code/watchman on git:5e7d4a0 ● [22:33:07]
    → make
    make  all-am
    make[1]: Entering directory '/home/russoue/code/watchman'
      CC       thirdparty/libart/src/art.o
    thirdparty/libart/src/art.c:4:23: fatal error: emmintrin.h: No such file or directory
     #include <emmintrin.h>
                           ^
    compilation terminated.
    Makefile:1263: recipe for target 'thirdparty/libart/src/art.o' failed
    make[1]: *** [thirdparty/libart/src/art.o] Error 1
    make[1]: Leaving directory '/home/russoue/code/watchman'
    Makefile:815: recipe for target 'all' failed
    make: *** [all] Error 2

From https://github.com/barrysteyn/node-scrypt/issues/44 it seems that it is because ARM does not support SSE2 instruction set. Can you support an ARM build?

**Comment by wez** [comment]
We don't have an arm build environment that we're going to support watchman on, but I think that the changes required are pretty small; if you don't mind hacking them together and trying it out and sending back a PR, we'll take it.

The piece of code that depends on the intrinsics is here:
https://github.com/facebook/watchman/blob/master/thirdparty/libart/src/art.c#L221-L237

I think it can be written as:

```
for (i=0;i < n->num_children; i++) {
      if (p.n16->keys[i] == c) {
           return &p.n16->children[i];
      }
}
break;
```

So what remains is amending configure.ac to `AC_CHECK_HEADER([emmintrin.h])` and then using the `#ifdef HAVE_EMMINTRIN_H` to control which of these is used.

You can run `make check` as a relatively quick check while you iterate on making it compile, then follow it up with `make integration` to make sure that is passes the test suite on your arm system.

**Comment by russoue** [comment]
Thanks, I applied those changes and found there are more dependencies on that header:

    make  all-am
    make[1]: Entering directory '/home/russoue/code/watchman'
      CC       thirdparty/libart/src/art.o
    thirdparty/libart/src/art.c: In function ‘find_child’:
    thirdparty/libart/src/art.c:211:18: warning: unused variable ‘bitfield’ [-Wunused-variable]
         int i, mask, bitfield;
                      ^
    thirdparty/libart/src/art.c:211:12: warning: unused variable ‘mask’ [-Wunused-variable]
         int i, mask, bitfield;
                ^
    thirdparty/libart/src/art.c: In function ‘add_child16’:
    thirdparty/libart/src/art.c:534:9: error: unknown type name ‘__m128i’
             __m128i cmp;
             ^
    thirdparty/libart/src/art.c:538:9: warning: implicit declaration of function ‘_mm_cmplt_epi8’ [-Wimplicit-function-declaration]
             cmp = _mm_cmplt_epi8(_mm_set1_epi8(c),
             ^
    thirdparty/libart/src/art.c:538:9: warning: implicit declaration of function ‘_mm_set1_epi8’ [-Wimplicit-function-declaration]
    thirdparty/libart/src/art.c:539:17: warning: implicit declaration of function ‘_mm_loadu_si128’ [-Wimplicit-function-declaration]
                     _mm_loadu_si128((__m128i*)n->keys));
                     ^
    thirdparty/libart/src/art.c:539:34: error: ‘__m128i’ undeclared (first use in this function)
                     _mm_loadu_si128((__m128i*)n->keys));
                                      ^
    thirdparty/libart/src/art.c:539:34: note: each undeclared identifier is reported only once for each function it appears in
    thirdparty/libart/src/art.c:539:42: error: expected expression before ‘)’ token
                     _mm_loadu_si128((__m128i*)n->keys));
                                              ^
    thirdparty/libart/src/art.c:543:9: warning: implicit declaration of function ‘_mm_movemask_epi8’ [-Wimplicit-function-declaration]
             bitfield = _mm_movemask_epi8(cmp) & mask;
             ^
    Makefile:1263: recipe for target 'thirdparty/libart/src/art.o' failed
    make[1]: *** [thirdparty/libart/src/art.o] Error 1
    make[1]: Leaving directory '/home/russoue/code/watchman'
    Makefile:815: recipe for target 'all' failed
    make: *** [all] Error 2

I will try to find if there are workarounds like your one for these too.

**Comment by wez** [comment]
I believe this to be fixed in master; would be great if you could confirm this!

**Comment by minikin** [comment]
I've just got the same error.

**Comment by wez** [comment]
@minikin did you try building from master?  If so, please share more details.

**Comment by minikin** [comment]
@wez I've just successfully built watchman from the master on raspberry pi3. Thanks!

**Comment by wez** [comment]
Thanks for confirming!  This will be released as watchman 4.7 in the next week or two

**Comment by wez** [comment]
v4.7.0 was tagged today!
